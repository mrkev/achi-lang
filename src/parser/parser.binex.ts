import * as Parsimmon from "parsimmon";
import { LangType } from "./parser";
import { Node, withAt } from "./Node";

type PrefixUnaryOperator = "-" | "!";
export type PrefixUnaryOperation = Node<{
  kind: "PrefixUnaryOperation";
  operator: PrefixUnaryOperator;
  value: NEXT_PARSER;
}>;

type SuffixUnaryOperator = "!";
export type SuffixUnaryOperation = Node<{
  kind: "SuffixUnaryOperation";
  operator: SuffixUnaryOperator;
  value: NEXT_PARSER;
}>;

type BinaryOperator =
  // String + Numerial
  | "+"
  // Numerical
  | "-"
  | "*"
  | "/"
  | "^"
  // Boolean
  | "&&"
  | "||"
  // Comparison
  | ">"
  | "<"
  | ">="
  | "<="
  | "=="
  | "!=";

export type BinaryOperation = Node<{
  kind: "BinaryOperation";
  operator: BinaryOperator;
  left: NEXT_PARSER;
  right: NEXT_PARSER;
}>;

export type OperatableExpression =
  | LangType["NumberLiteral"]
  | LangType["ValueIdentifier"]
  | LangType["StringLiteral"]
  | LangType["BooleanLiteral"]
  | LangType["FunctionCall"]
  | LangType["NullLiteral"];

// This parser supports basic math with + - * / ^, unary negation, factorial,
// and parentheses. It does not evaluate the math, just turn it into a series of
// nested lists that are easy to evaluate.

// You might think that parsing math would be easy since people learn it early
// in school, but dealing with precedence and associativity of operators is
// actually one of the hardest and most tedious things you can do in a parser!
// If you look at a language like JavaScript, it has even more operators than
// math, like = and && and || and ++ and so many more...

///////////////////////////////////////////////////////////////////////

type NEXT_PARSER =
  | OperatableExpression
  | PrefixUnaryOperation
  | SuffixUnaryOperation
  | BinaryOperation;

// Takes a parser for the prefix operator, and a parser for the base thing being
// parsed, and parses as many occurrences as possible of the prefix operator.
// Note that the parser is created using `Parsimmon.lazy` because it's recursive. It's
// valid for there to be zero occurrences of the prefix operator.
function PREFIX(
  operatorsParser: Parsimmon.Parser<OpLoc<PrefixUnaryOperator>>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
): Parsimmon.Parser<NEXT_PARSER> {
  const parser = Parsimmon.lazy<NEXT_PARSER>(() => {
    const res: Parsimmon.Parser<NEXT_PARSER> = Parsimmon.seqMap(
      Parsimmon.index,
      operatorsParser,
      parser,
      Parsimmon.index,
      (start, operator, value, end) =>
        ({
          kind: "PrefixUnaryOperation",
          operator: operator.op,
          value,
          "@": { start, end },
        } as const)
    ).or(nextParser);
    return res;
  });
  return parser;
}

type OpLoc<T extends string> = {
  op: T;
  start: Parsimmon.Index;
  end: Parsimmon.Index;
};

// Ideally this function would be just like `PREFIX` but reordered like
// `Parsimmon.seq(parser, operatorsParser).or(nextParser)`, but that doesn't work. The
// reason for that is that Parsimmon will get stuck in infinite recursion, since
// the very first rule. Inside `parser` is to match parser again. Alternatively,
// you might think to try `nextParser.or(Parsimmon.seq(parser, operatorsParser))`, but
// that won't work either because in a call to `.or` (aka `Parsimmon.alt`), Parsimmon
// takes the first possible match, even if subsequent matches are longer, so the
// parser will never actually look far enough ahead to see the postfix
// operators.
function POSTFIX(
  operatorsParser: Parsimmon.Parser<OpLoc<"!">>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
) {
  // Because we can't use recursion like stated above, we just match a flat list
  // of as many occurrences of the postfix operator as possible, then use
  // `.reduce` to manually nest the list.
  //
  // Example:
  //
  // INPUT  :: "4!!!"
  // PARSE  :: [4, "factorial", "factorial", "factorial"]
  // REDUCE :: ["factorial", ["factorial", ["factorial", 4]]]
  return Parsimmon.seqMap(nextParser, operatorsParser.many(), (x, suffixes) =>
    suffixes.reduce((acc, { op, end }) => {
      return {
        kind: "SuffixUnaryOperation",
        operator: op,
        value: acc,
        "@": {
          start: acc["@"].start,
          end,
        },
      } as const;
    }, x)
  );
}

// Takes a parser for all the operators at this precedence level, and a parser
// that parsers everything at the next precedence level, and returns a parser
// that parses as many binary operations as possible, associating them to the
// right. (e.g. 1^2^3 is 1^(2^3) not (1^2)^3)
function BINARY_RIGHT(
  operatorsParser: Parsimmon.Parser<OpLoc<BinaryOperator>>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
): Parsimmon.Parser<NEXT_PARSER> {
  const parser = Parsimmon.lazy<NEXT_PARSER>(() => {
    const chainRes: Parsimmon.Parser<NEXT_PARSER> = nextParser.chain((next) =>
      withAt(
        Parsimmon.seqMap(
          operatorsParser,
          Parsimmon.of(next),
          parser,
          (operator, next, p) =>
            ({
              kind: "BinaryOperation",
              operator: operator.op,
              left: next,
              right: p,
            } as const)
        )
      ).or(Parsimmon.of(next))
    );

    return chainRes;
  });
  return parser;
}

// Takes a parser for all the operators at this precedence level, and a parser
// that parsers everything at the next precedence level, and returns a parser
// that parses as many binary operations as possible, associating them to the
// left. (e.g. 1-2-3 is (1-2)-3 not 1-(2-3))
function BINARY_LEFT(
  operatorsParser: Parsimmon.Parser<OpLoc<BinaryOperator>>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
) {
  // We run into a similar problem as with the `POSTFIX` parser above where we
  // can't recurse in the direction we want, so we have to resort to parsing an
  // entire list of operator chunks and then using `.reduce` to manually nest
  // them again.
  //
  // Example:
  //
  // INPUT  :: "1+2+3"
  // PARSE  :: [1, ["+", 2], ["+", 3]]
  // REDUCE :: ["+", ["+", 1, 2], 3]
  return Parsimmon.seqMap(
    nextParser,
    Parsimmon.seq(operatorsParser, nextParser).many(),
    (first, rest) => {
      return rest.reduce((acc, ch) => {
        const [op, another] = ch;
        return {
          kind: "BinaryOperation",
          operator: op.op,
          left: acc,
          right: another,
          "@": {
            start: acc["@"].start,
            end: another["@"].end,
          },
        } as const;
      }, first);
    }
  );
}

export function OperatorParser(r: Parsimmon.TypedLanguage<LangType>) {
  // Operators should allow whitespace around them, but not require it. This
  // helper combines multiple operators together with names.
  //
  // Example: operators({Add: "+", Sub: "-"})
  //
  // Gives back an operator that parses either + or - surrounded by optional
  // whitespace, and gives back the word "Add" or "Sub" instead of the character.
  function operators<Ops extends readonly string[]>(
    ops: Ops
  ): Parsimmon.Parser<OpLoc<Ops[number]>> {
    const withWhitespace = ops.map((sym) =>
      Parsimmon.seqMap(
        Parsimmon.index,
        r._,
        Parsimmon.string(sym),
        r._,
        Parsimmon.index,
        (start, _0, op, _2, end) => {
          return { op, start, end };
        }
      )
    );
    const res = Parsimmon.alt(...withWhitespace);
    return res;
  }

  // A basic value is any parenthesized expression or a number.
  // TODO: start, end don't take the parens into consideration. Is that an issue?
  const Basic: Parsimmon.Parser<NEXT_PARSER> = Parsimmon.lazy<NEXT_PARSER>(() =>
    Parsimmon.string("(")
      .then(MyMath)
      .skip(Parsimmon.string(")"))
      .or(r.NumberLiteral)
      .or(r.StringLiteral)
      .or(r.BooleanLiteral)
      .or(r.FunctionCall)
      .or(r.NullLiteral)
      .or(r.ValueIdentifier)
  );

  // Now we can describe the operators in order by precedence. You just need to
  // re-order the table.
  // const table = [
  //   { type: POSTFIX, ops: operators(["!"] as const) },
  //   { type: PREFIX, ops: operators(["-", "!"] as const) },
  //   { type: BINARY_RIGHT, ops: operators(["^"] as const) },
  //   {
  //     type: BINARY_LEFT,
  //     ops: operators(["*", "/"] as const),
  //   },
  //   {
  //     type: BINARY_LEFT,
  //     ops: operators(["+", "-"] as const),
  //   },
  // ];

  const table: Array<
    (next: Parsimmon.Parser<NEXT_PARSER>) => Parsimmon.Parser<NEXT_PARSER>
  > = [
    (next) => BINARY_LEFT(operators(["==", "!="] as const), next),
    (next) => POSTFIX(operators(["!"] as const), next),
    (next) => PREFIX(operators(["-", "!"] as const), next),
    (next) => BINARY_RIGHT(operators(["^"] as const), next),
    (next) => BINARY_LEFT(operators(["*", "/"] as const), next),
    (next) => BINARY_LEFT(operators(["+", "-"] as const), next),
    (next) => BINARY_LEFT(operators(["&&"] as const), next),
    (next) => BINARY_LEFT(operators(["||"] as const), next),
    (next) => BINARY_LEFT(operators([">=", "<=", ">", "<"] as const), next),
  ];

  // Start off with Num as the base parser for numbers and thread that through the
  // entire table of operator parsers.
  const tableParser = table.reduce((next, level) => level(next), Basic);

  // The above is equivalent to:
  //
  // TYPE(operators({...}),
  //   TYPE(operators({...}),
  //     TYPE(operators({...})),
  //       TYPE(operators({...}),
  //         TYPE(operators({...}), ...))))
  //
  // But it's easier if to see what's going on and reorder the precedence if we
  // keep it in a table instead of nesting it all manually.

  // This is our version of a math expression.
  const MyMath = tableParser.trim(r._);
  return MyMath;
}
