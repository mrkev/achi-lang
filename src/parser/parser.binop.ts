import * as Parsimmon from "parsimmon";
import { LangType } from "./parser";
import { MathParser, NEXT_PARSER } from "./parser.binex";
import { sublang } from "./sublang";

// expression =
//     term
//     | expression "+" term
//     | expression "-" term .
// term =
//     factor
//     | term "*" factor
//     | term "/" factor .
// factor =
//     number
//     | "(" expression ")" .

export type BinaryExpression = {
  kind: "BinaryExpression";
  operator: string;
  left: LangType["Expression"];
  right: LangType["Expression"];
};

export type LangType_BinOp = {
  // BinaryOperator: string;
  // BinaryOperation: {
  //   kind: "BinaryOperation";
  //   operator: string;
  //   left: LangType["Expression"];
  //   right: LangType["Expression"];
  // };

  // BinaryExpression_Expr
  // BinaryExpression: BinaryExpression | LangType["Expression"];
  // BinaryExpression_Term: BinaryExpression | LangType["Expression"];
  // BinaryExpression_Factor: BinaryExpression | LangType["Expression"];
  OperatorExpression: NEXT_PARSER;
};

/**
 * Defines binary operations
 */
export const LangDef_BinOp = sublang<LangType, LangType_BinOp>({
  OperatorExpression: MathParser,

  // BinaryExpression: (r) => {
  //   return Parsimmon.alt(
  //     r.BinaryExpression_Term,
  //     Parsimmon.seqMap(
  //       r.BinaryExpression,
  //       r._,
  //       Parsimmon.regex(/(\+|\-)/),
  //       r._,
  //       r.BinaryExpression_Term,
  //       (expr, _1, op, _3, term): LangType["BinaryExpression"] => {
  //         return {
  //           kind: "BinaryExpression",
  //           left: expr,
  //           operator: op,
  //           right: term,
  //         };
  //       }
  //     )
  //   );
  // },

  // // // term =
  // //     factor
  // //     | term "*" factor
  // //     | term "/" factor .
  // BinaryExpression_Term: (r) => {
  //   return Parsimmon.alt(
  //     r.BinaryExpression_Factor,
  //     Parsimmon.seqMap(
  //       r.BinaryExpression_Term,
  //       r._,
  //       Parsimmon.regex(/(\*|\/)/),
  //       r._,
  //       r.BinaryExpression_Factor,
  //       (term, _1, op, _3, factor): LangType["BinaryExpression_Term"] => {
  //         return {
  //           kind: "BinaryExpression",
  //           left: term,
  //           operator: op,
  //           right: factor,
  //         };
  //       }
  //     )
  //   );
  // },

  // BinaryExpression_Factor: (r) => {
  //   return Parsimmon.seqMap(
  //     Parsimmon.string("<<"),
  //     r._,
  //     r.BinaryExpression,
  //     r._,
  //     Parsimmon.string(">>"),
  //     (_0, _1, expr, _3, _4) => {
  //       return expr;
  //     }
  //   ).or(r.NumberLiteral);
  // },

  // Arithmetic: + - / * %
  // Comparison: all double << >> == !=
  // Logic: || &&
  // Todo, bitwise?
  // BinaryOperator: () => {
  //   return Parsimmon.regex(/(\+|\-|\/|\*|\%|\=\=|\!\=|\>\>|\<\<|\|\||\&\&)/);
  // },

  // 3 + 3
  // BinaryOperation: (r) => {
  //   return Parsimmon.seqMap(
  //     r.Expression,
  //     r._,
  //     r.BinaryOperator,
  //     r._,
  //     r.Expression,
  //     (left, _0, operator, _1, right) => {
  //       return {
  //         kind: "BinaryOperation",
  //         operator,
  //         left,
  //         right,
  //       };
  //     }
  //   );
  // },
});

// Factor
// = "(" _ expr:Expression _ ")" { return expr; }
// / Integer
