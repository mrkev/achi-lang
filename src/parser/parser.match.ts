import * as Parsimmon from "parsimmon";
import { LangType } from "./parser";
import { sublang } from "./sublang";

export type LangType_Match = {
  CaseEntry: {
    kind: "CaseEntry";
    guard: LangType["Expression"];
    block: LangType["Block"];
  };

  MatchFunction: {
    kind: "MatchFunction";
    identifier: LangType["ValueIdentifier"];
    matchType: LangType["Type"];
    block: LangType["BlockOfCases"];
  };

  BlockOfCases: {
    kind: "BlockOfCases";
    caseEntries: Array<LangType["CaseEntry"]>;
  };
};

/**
 * Defines binary operations
 */
export const LangDef_Match = sublang<LangType, LangType_Match>({
  // case (card: King(value)): { ... }
  CaseEntry: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("case"),
      r.__,
      r.Expression,
      r._,
      Parsimmon.string(":"),
      r.__,
      r.Block,
      function (_0, _1, expression, _3, _4, _5, block) {
        return {
          kind: "CaseEntry",
          guard: expression,
          block,
        };
      }
    );
  },

  // function foo matches Type {}
  MatchFunction: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("function"),
      r.__,
      r.ValueIdentifier,
      r.__,
      Parsimmon.string("matches"),
      r.__,
      r.Type,
      r.__,
      r.BlockOfCases,
      function (_0, _1, identifier, _3, _4, _5, type, _7, block) {
        return {
          kind: "MatchFunction",
          identifier,
          matchType: type,
          block: block,
        };
      }
    );
  },

  BlockOfCases: (r) => {
    const caseListParser = Parsimmon.sepBy(r.CaseEntry, r.__nl).trim(r._nl);

    return (
      Parsimmon.string("{")
        // TBD: sepBy1 and make a different parser for Unit?
        .then(caseListParser)
        .skip(Parsimmon.string("}"))
        .map((caseEntries) => {
          return {
            kind: "BlockOfCases",
            caseEntries,
          };
        })
    );
  },
});
