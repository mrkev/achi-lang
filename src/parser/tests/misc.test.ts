import { Lang } from "../parser";

test("Program", () => {
  const result = Lang.Program.tryParse(`
  class Point(x: number, y: number);
  const one = 1;
  const point = Point(x: one, y: 2);

  #log point
`);
  expect(result).toMatchSnapshot();
});

test("Program.2", () => {
  const result = Lang.Program.tryParse(`
  class Point(x: number, y: number);
  const increment = (point: Point) => {
    const x = point;
    return Point(x: x, y: y)
  };
  const point = Point(x: one, y: 2);
  increment(point: point);

  #log point
`);
  expect(result).toMatchSnapshot();
});

test("ReturnStatement", () => {
  const result = Lang.ReturnStatement.tryParse(`return "string"`);
  expect(result).toMatchSnapshot();
});

test("Block.empty", () => {
  const result = Lang.Block.tryParse(`{}`);
  expect(result).toMatchSnapshot();
});

test("IfStatement", () => {
  const result = Lang.IfStatement.tryParse(`if (3) { return "hello" }`);
  expect(result).toMatchSnapshot();
});

test("NestedTypeIdentifier", () => {
  const result = Lang.NestedTypeIdentifier.tryParse("Card.King");
  expect(result).toMatchInlineSnapshot(`
{
  "kind": "NestedTypeIdentifier",
  "path": [
    {
      "kind": "TypeIdentifier",
      "value": "Card",
    },
    {
      "kind": "TypeIdentifier",
      "value": "King",
    },
  ],
}
`);
});

test("ConstantDefinition", () => {
  const result = Lang.ConstantDefinition.tryParse("const x = 3");
  expect(result).toMatchSnapshot();
});

test("ConstantDefinition.functionCall", () => {
  const result = Lang.ConstantDefinition.tryParse(
    "const x = foo(point: point)"
  );
  expect(result).toMatchSnapshot();
});

test("Block", () => {
  const result = Lang.Block.tryParse(`{ log(msg: "hello"); const x = 4; }`);
  expect(result).toMatchSnapshot();
});

test("string", () => {
  expect(Lang.StringLiteral.tryParse(`"hello world"`)).toMatchSnapshot();
  expect(Lang.StringLiteral.tryParse(`"hello\\n world"`)).toMatchSnapshot();
});

// Types

// test("NamedTupleDefinition", () => {
//   const result = Lang.NamedTupleDefinition.tryParse(
//     "type Point(number, number)"
//   );
//   expect(result).toMatchSnapshot();
// });

// test("TupleDefinition", () => {
//   const result = Lang.TupleDefinition.tryParse("(number, number)");
//   expect(result).toMatchSnapshot();
// });

// -------------- Atoms
