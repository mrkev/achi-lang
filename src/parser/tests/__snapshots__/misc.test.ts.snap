// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block 1`] = `
{
  "kind": "Block",
  "statements": [
    {
      "argument": {
        "definitions": [
          {
            "expression": {
              "kind": "StringLiteral",
              "value": "hello",
            },
            "identifier": {
              "_meta": {
                "end": {
                  "column": 10,
                  "line": 1,
                  "offset": 9,
                },
                "start": {
                  "column": 7,
                  "line": 1,
                  "offset": 6,
                },
              },
              "kind": "ValueIdentifier",
              "value": "msg",
            },
            "kind": "NamedLiteral",
          },
        ],
        "kind": "RecordLiteral",
      },
      "identifier": {
        "_meta": {
          "end": {
            "column": 6,
            "line": 1,
            "offset": 5,
          },
          "start": {
            "column": 3,
            "line": 1,
            "offset": 2,
          },
        },
        "kind": "ValueIdentifier",
        "value": "log",
      },
      "kind": "FunctionCall",
    },
    {
      "expression": {
        "kind": "NumberLiteral",
        "value": 4,
      },
      "identifier": {
        "_meta": {
          "end": {
            "column": 29,
            "line": 1,
            "offset": 28,
          },
          "start": {
            "column": 28,
            "line": 1,
            "offset": 27,
          },
        },
        "kind": "ValueIdentifier",
        "value": "x",
      },
      "kind": "ConstantDefinition",
    },
  ],
}
`;

exports[`Block.empty 1`] = `
{
  "kind": "Block",
  "statements": [],
}
`;

exports[`ConstantDefinition 1`] = `
{
  "expression": {
    "kind": "NumberLiteral",
    "value": 3,
  },
  "identifier": {
    "_meta": {
      "end": {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
      "start": {
        "column": 7,
        "line": 1,
        "offset": 6,
      },
    },
    "kind": "ValueIdentifier",
    "value": "x",
  },
  "kind": "ConstantDefinition",
}
`;

exports[`ConstantDefinition.functionCall 1`] = `
{
  "expression": {
    "argument": {
      "definitions": [
        {
          "expression": {
            "_meta": {
              "end": {
                "column": 27,
                "line": 1,
                "offset": 26,
              },
              "start": {
                "column": 22,
                "line": 1,
                "offset": 21,
              },
            },
            "kind": "ValueIdentifier",
            "value": "point",
          },
          "identifier": {
            "_meta": {
              "end": {
                "column": 20,
                "line": 1,
                "offset": 19,
              },
              "start": {
                "column": 15,
                "line": 1,
                "offset": 14,
              },
            },
            "kind": "ValueIdentifier",
            "value": "point",
          },
          "kind": "NamedLiteral",
        },
      ],
      "kind": "RecordLiteral",
    },
    "identifier": {
      "_meta": {
        "end": {
          "column": 14,
          "line": 1,
          "offset": 13,
        },
        "start": {
          "column": 11,
          "line": 1,
          "offset": 10,
        },
      },
      "kind": "ValueIdentifier",
      "value": "foo",
    },
    "kind": "FunctionCall",
  },
  "identifier": {
    "_meta": {
      "end": {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
      "start": {
        "column": 7,
        "line": 1,
        "offset": 6,
      },
    },
    "kind": "ValueIdentifier",
    "value": "x",
  },
  "kind": "ConstantDefinition",
}
`;

exports[`IfStatement 1`] = `
{
  "block": {
    "kind": "Block",
    "statements": [
      {
        "expression": {
          "kind": "StringLiteral",
          "value": "hello",
        },
        "kind": "ReturnStatement",
      },
    ],
  },
  "guard": {
    "kind": "NumberLiteral",
    "value": 3,
  },
  "kind": "IfStatement",
}
`;

exports[`Program 1`] = `
{
  "kind": "Program",
  "statements": [
    {
      "kind": "NamedRecordDefinitionStatement",
      "namedRecordDefinition": {
        "identifier": {
          "kind": "TypeIdentifier",
          "value": "Point",
        },
        "kind": "NamedRecordDefinition",
        "record": {
          "definitions": [
            {
              "identifier": {
                "_meta": {
                  "end": {
                    "column": 16,
                    "line": 2,
                    "offset": 16,
                  },
                  "start": {
                    "column": 15,
                    "line": 2,
                    "offset": 15,
                  },
                },
                "kind": "ValueIdentifier",
                "value": "x",
              },
              "kind": "NamedDefinition",
              "typeTag": {
                "identifier": {
                  "kind": "TypeIdentifier",
                  "value": "number",
                },
                "kind": "TypeTag",
              },
            },
            {
              "identifier": {
                "_meta": {
                  "end": {
                    "column": 27,
                    "line": 2,
                    "offset": 27,
                  },
                  "start": {
                    "column": 26,
                    "line": 2,
                    "offset": 26,
                  },
                },
                "kind": "ValueIdentifier",
                "value": "y",
              },
              "kind": "NamedDefinition",
              "typeTag": {
                "identifier": {
                  "kind": "TypeIdentifier",
                  "value": "number",
                },
                "kind": "TypeTag",
              },
            },
          ],
          "kind": "RecordDefinition",
        },
      },
    },
    {
      "expression": {
        "kind": "NumberLiteral",
        "value": 1,
      },
      "identifier": {
        "_meta": {
          "end": {
            "column": 12,
            "line": 3,
            "offset": 49,
          },
          "start": {
            "column": 9,
            "line": 3,
            "offset": 46,
          },
        },
        "kind": "ValueIdentifier",
        "value": "one",
      },
      "kind": "ConstantDefinition",
    },
    {
      "expression": {
        "identifier": {
          "kind": "TypeIdentifier",
          "value": "Point",
        },
        "kind": "NamedRecordLiteral",
        "recordLiteral": {
          "definitions": [
            {
              "expression": {
                "_meta": {
                  "end": {
                    "column": 29,
                    "line": 4,
                    "offset": 83,
                  },
                  "start": {
                    "column": 26,
                    "line": 4,
                    "offset": 80,
                  },
                },
                "kind": "ValueIdentifier",
                "value": "one",
              },
              "identifier": {
                "_meta": {
                  "end": {
                    "column": 24,
                    "line": 4,
                    "offset": 78,
                  },
                  "start": {
                    "column": 23,
                    "line": 4,
                    "offset": 77,
                  },
                },
                "kind": "ValueIdentifier",
                "value": "x",
              },
              "kind": "NamedLiteral",
            },
            {
              "expression": {
                "kind": "NumberLiteral",
                "value": 2,
              },
              "identifier": {
                "_meta": {
                  "end": {
                    "column": 32,
                    "line": 4,
                    "offset": 86,
                  },
                  "start": {
                    "column": 31,
                    "line": 4,
                    "offset": 85,
                  },
                },
                "kind": "ValueIdentifier",
                "value": "y",
              },
              "kind": "NamedLiteral",
            },
          ],
          "kind": "RecordLiteral",
        },
      },
      "identifier": {
        "_meta": {
          "end": {
            "column": 14,
            "line": 4,
            "offset": 68,
          },
          "start": {
            "column": 9,
            "line": 4,
            "offset": 63,
          },
        },
        "kind": "ValueIdentifier",
        "value": "point",
      },
      "kind": "ConstantDefinition",
    },
    {
      "expression": {
        "_meta": {
          "end": {
            "column": 0,
            "line": 7,
            "offset": 105,
          },
          "start": {
            "column": 8,
            "line": 6,
            "offset": 100,
          },
        },
        "kind": "ValueIdentifier",
        "value": "point",
      },
      "kind": "DEBUG_Log",
    },
  ],
}
`;

exports[`Program.2 1`] = `
{
  "kind": "Program",
  "statements": [
    {
      "kind": "NamedRecordDefinitionStatement",
      "namedRecordDefinition": {
        "identifier": {
          "kind": "TypeIdentifier",
          "value": "Point",
        },
        "kind": "NamedRecordDefinition",
        "record": {
          "definitions": [
            {
              "identifier": {
                "_meta": {
                  "end": {
                    "column": 16,
                    "line": 2,
                    "offset": 16,
                  },
                  "start": {
                    "column": 15,
                    "line": 2,
                    "offset": 15,
                  },
                },
                "kind": "ValueIdentifier",
                "value": "x",
              },
              "kind": "NamedDefinition",
              "typeTag": {
                "identifier": {
                  "kind": "TypeIdentifier",
                  "value": "number",
                },
                "kind": "TypeTag",
              },
            },
            {
              "identifier": {
                "_meta": {
                  "end": {
                    "column": 27,
                    "line": 2,
                    "offset": 27,
                  },
                  "start": {
                    "column": 26,
                    "line": 2,
                    "offset": 26,
                  },
                },
                "kind": "ValueIdentifier",
                "value": "y",
              },
              "kind": "NamedDefinition",
              "typeTag": {
                "identifier": {
                  "kind": "TypeIdentifier",
                  "value": "number",
                },
                "kind": "TypeTag",
              },
            },
          ],
          "kind": "RecordDefinition",
        },
      },
    },
    {
      "expression": {
        "argument": {
          "definitions": [
            {
              "identifier": {
                "_meta": {
                  "end": {
                    "column": 27,
                    "line": 3,
                    "offset": 64,
                  },
                  "start": {
                    "column": 22,
                    "line": 3,
                    "offset": 59,
                  },
                },
                "kind": "ValueIdentifier",
                "value": "point",
              },
              "kind": "NamedDefinition",
              "typeTag": {
                "identifier": {
                  "kind": "TypeIdentifier",
                  "value": "Point",
                },
                "kind": "TypeTag",
              },
            },
          ],
          "kind": "RecordDefinition",
        },
        "block": {
          "kind": "Block",
          "statements": [
            {
              "expression": {
                "_meta": {
                  "end": {
                    "column": 20,
                    "line": 4,
                    "offset": 97,
                  },
                  "start": {
                    "column": 15,
                    "line": 4,
                    "offset": 92,
                  },
                },
                "kind": "ValueIdentifier",
                "value": "point",
              },
              "identifier": {
                "_meta": {
                  "end": {
                    "column": 12,
                    "line": 4,
                    "offset": 89,
                  },
                  "start": {
                    "column": 11,
                    "line": 4,
                    "offset": 88,
                  },
                },
                "kind": "ValueIdentifier",
                "value": "x",
              },
              "kind": "ConstantDefinition",
            },
            {
              "expression": {
                "identifier": {
                  "kind": "TypeIdentifier",
                  "value": "Point",
                },
                "kind": "NamedRecordLiteral",
                "recordLiteral": {
                  "definitions": [
                    {
                      "expression": {
                        "_meta": {
                          "end": {
                            "column": 22,
                            "line": 5,
                            "offset": 120,
                          },
                          "start": {
                            "column": 21,
                            "line": 5,
                            "offset": 119,
                          },
                        },
                        "kind": "ValueIdentifier",
                        "value": "x",
                      },
                      "identifier": {
                        "_meta": {
                          "end": {
                            "column": 19,
                            "line": 5,
                            "offset": 117,
                          },
                          "start": {
                            "column": 18,
                            "line": 5,
                            "offset": 116,
                          },
                        },
                        "kind": "ValueIdentifier",
                        "value": "x",
                      },
                      "kind": "NamedLiteral",
                    },
                    {
                      "expression": {
                        "_meta": {
                          "end": {
                            "column": 28,
                            "line": 5,
                            "offset": 126,
                          },
                          "start": {
                            "column": 27,
                            "line": 5,
                            "offset": 125,
                          },
                        },
                        "kind": "ValueIdentifier",
                        "value": "y",
                      },
                      "identifier": {
                        "_meta": {
                          "end": {
                            "column": 25,
                            "line": 5,
                            "offset": 123,
                          },
                          "start": {
                            "column": 24,
                            "line": 5,
                            "offset": 122,
                          },
                        },
                        "kind": "ValueIdentifier",
                        "value": "y",
                      },
                      "kind": "NamedLiteral",
                    },
                  ],
                  "kind": "RecordLiteral",
                },
              },
              "kind": "ReturnStatement",
            },
          ],
        },
        "kind": "AnonymousFunctionLiteral",
      },
      "identifier": {
        "_meta": {
          "end": {
            "column": 18,
            "line": 3,
            "offset": 55,
          },
          "start": {
            "column": 9,
            "line": 3,
            "offset": 46,
          },
        },
        "kind": "ValueIdentifier",
        "value": "increment",
      },
      "kind": "ConstantDefinition",
    },
    {
      "expression": {
        "identifier": {
          "kind": "TypeIdentifier",
          "value": "Point",
        },
        "kind": "NamedRecordLiteral",
        "recordLiteral": {
          "definitions": [
            {
              "expression": {
                "_meta": {
                  "end": {
                    "column": 29,
                    "line": 7,
                    "offset": 161,
                  },
                  "start": {
                    "column": 26,
                    "line": 7,
                    "offset": 158,
                  },
                },
                "kind": "ValueIdentifier",
                "value": "one",
              },
              "identifier": {
                "_meta": {
                  "end": {
                    "column": 24,
                    "line": 7,
                    "offset": 156,
                  },
                  "start": {
                    "column": 23,
                    "line": 7,
                    "offset": 155,
                  },
                },
                "kind": "ValueIdentifier",
                "value": "x",
              },
              "kind": "NamedLiteral",
            },
            {
              "expression": {
                "kind": "NumberLiteral",
                "value": 2,
              },
              "identifier": {
                "_meta": {
                  "end": {
                    "column": 32,
                    "line": 7,
                    "offset": 164,
                  },
                  "start": {
                    "column": 31,
                    "line": 7,
                    "offset": 163,
                  },
                },
                "kind": "ValueIdentifier",
                "value": "y",
              },
              "kind": "NamedLiteral",
            },
          ],
          "kind": "RecordLiteral",
        },
      },
      "identifier": {
        "_meta": {
          "end": {
            "column": 14,
            "line": 7,
            "offset": 146,
          },
          "start": {
            "column": 9,
            "line": 7,
            "offset": 141,
          },
        },
        "kind": "ValueIdentifier",
        "value": "point",
      },
      "kind": "ConstantDefinition",
    },
    {
      "argument": {
        "definitions": [
          {
            "expression": {
              "_meta": {
                "end": {
                  "column": 25,
                  "line": 8,
                  "offset": 194,
                },
                "start": {
                  "column": 20,
                  "line": 8,
                  "offset": 189,
                },
              },
              "kind": "ValueIdentifier",
              "value": "point",
            },
            "identifier": {
              "_meta": {
                "end": {
                  "column": 18,
                  "line": 8,
                  "offset": 187,
                },
                "start": {
                  "column": 13,
                  "line": 8,
                  "offset": 182,
                },
              },
              "kind": "ValueIdentifier",
              "value": "point",
            },
            "kind": "NamedLiteral",
          },
        ],
        "kind": "RecordLiteral",
      },
      "identifier": {
        "_meta": {
          "end": {
            "column": 12,
            "line": 8,
            "offset": 181,
          },
          "start": {
            "column": 3,
            "line": 8,
            "offset": 172,
          },
        },
        "kind": "ValueIdentifier",
        "value": "increment",
      },
      "kind": "FunctionCall",
    },
    {
      "expression": {
        "_meta": {
          "end": {
            "column": 0,
            "line": 11,
            "offset": 210,
          },
          "start": {
            "column": 8,
            "line": 10,
            "offset": 205,
          },
        },
        "kind": "ValueIdentifier",
        "value": "point",
      },
      "kind": "DEBUG_Log",
    },
  ],
}
`;

exports[`ReturnStatement 1`] = `
{
  "expression": {
    "kind": "StringLiteral",
    "value": "string",
  },
  "kind": "ReturnStatement",
}
`;

exports[`string 1`] = `
{
  "kind": "StringLiteral",
  "value": "hello world",
}
`;

exports[`string 2`] = `
{
  "kind": "StringLiteral",
  "value": "hello
 world",
}
`;
